### PR TITLE
metrics-generator: include messaging systems and databases in service graphs

### PR DESCRIPTION
**Pull request by kvrhdn**
**What this PR does**:
Include databases and requests across messaging systems in service graphs metrics. We rely on the OpenTelemetry semantic conventions to detect when a span is involved in one of these systems.

Links:
- messaging system: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md
- databases: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md

**Messaging system**: span kind will be `consumer` and `producer`. We add `connection_type="messaging_system"` to the metrics.

**Database**: span kind will be `client` but there will be no matching `server` span (the database is most likely not instrumented). We copy the values from the client side and use that to populate the server latency. Downside: you can't track the client and server-side latency of a request.
We add `connection_type="database"` to the metrics.

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/tempo/issues/1435

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by kvrhdn** [comment]
For testing: you can run this Tempo image in the intro-to-mlt demo https://github.com/grafana/intro-to-mlt/tree/kvrhdn/add-rabbitmq
I'll add proper test cases soon.

This will generate the following connections:
Screenshot 2022-07-14 at 19 38 49

File: modules/generator/processor/servicegraphs/store/store.go [hunk 1]
```diff
@@ -93,16 +93,22 @@ func (s *store) UpsertEdge(key string, update Callback) (bool, error) {
...
 	}
 
+	edge := newEdge(key, s.ttl)
+	update(edge)
+
+	if edge.isComplete() {
+		s.onComplete(edge)
+		return true, nil
+	}
+
```
Comment (mapno) [inline modules/generator/processor/servicegraphs/store/store.go:103]: Nice improvement 👍

**Review by mapno — COMMENTED** [review]
Don't have a lot to comment. It looks good 👍

**Comment by kvrhdn** [comment]
I'll update the corresponding docs in a separate PR since I've a couple of changes pending and it's just going to clutter this PR.

File: modules/generator/processor/servicegraphs/servicegraphs.go [hunk 2]
```diff
@@ -149,20 +149,41 @@ func (p *Processor) consume(resourceSpans []*v1_trace.ResourceSpans) (err error)
 
 		for _, ils := range rs.InstrumentationLibrarySpans {
 			for _, span := range ils.Spans {
+				connectionType := store.Unknown
+
 				switch span.Kind {
+				case v1_trace.Span_SPAN_KIND_PRODUCER:
+					connectionType = store.MessagingSystem
+					fallthrough
```
Comment (annanay25) [inline modules/generator/processor/servicegraphs/servicegraphs.go:157]: I think we should add a comment here to emphasise why the producer case was placed before client and similarly for consumer <> server.

**Review by annanay25 — APPROVED** [review]

File: modules/generator/processor/servicegraphs/servicegraphs.go (hunk 2 above)
Comment (zalegrala) [reply modules/generator/processor/servicegraphs/servicegraphs.go:157]: Whoa, I've never seen `fallthrough` before.  Neat.

**Review by zalegrala — APPROVED** [review]

**Pull request merged**